File: service/shared/models/conversation.go
package models

import (
	"encoding/base64"
	"encoding/json"
	"time"
)

type Conversation struct {
	ID           int       `json:"conversation_id"`
	Name         string    `json:"name"`
	Photo        []byte    `json:"-"`
	Is_group     bool      `json:"is_group"`
	Created_at   time.Time `json:"created_at"`
	Messages     []Message `json:"messages"`
	Participants []User    `json:"participants"`
}

func (c *Conversation) MarshalJSON() ([]byte, error) {
	type Alias Conversation
	var photoStr string
	if len(c.Photo) > 0 {
		photoStr = base64.StdEncoding.EncodeToString(c.Photo)
	}

	return json.Marshal(&struct {
		*Alias
		Photo string `json:"photo,omitempty"`
	}{
		Alias: (*Alias)(c),
		Photo: photoStr,
	})
}

File: webui/src/modules/conversation/models/conversation.ts
import { Message, type MessageDTO } from '@/modules/message/models/message'
import { ConversationParticipant } from './conversation_participant'

export class Conversation {
  public conversationId: number
  public name: string
  public isGroup: boolean
  public createdAt: Date
  public participants: ConversationParticipant[]
  public messages?: Message[]

  constructor(data: ConversationDTO) {
    this.conversationId = data.conversation_id
    this.name = data.name
    this.isGroup = data.is_group
    this.createdAt = new Date(data.created_at)
    this.participants = data.conversation_participants
    this.messages = data.messages ? data.messages.map((messageDto) => new Message(messageDto)) : []
  }

  static fromJSON(json: ConversationDTO): Conversation {
    return new Conversation(json)
  }

  toJSON(): ConversationDTO {
    return {
      conversation_id: this.conversationId,
      name: this.name,
      is_group: this.isGroup,
      created_at: this.createdAt.toISOString(),
      conversation_participants: this.participants,
      messages: this.messages?.map((message) => message.toJSON()) || [],
    }
  }
}

export interface ConversationDTO {
  conversation_id: number
  name: string
  is_group: boolean
  created_at: string
  conversation_participants: ConversationParticipant[]
  messages?: MessageDTO[]
}
